### PR TITLE
Support auto_select_gpus with accelerator and devices api

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Supports #12590 

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/properties/test_auto_gpu_select.py [hunk 1]
```diff
@@ -52,10 +52,30 @@ def test_pick_multiple_gpus(nb, expected_gpu_idxs, expected_error):
...
             ),
         ):
             pick_multiple_gpus(nb)
     else:
         assert expected_gpu_idxs == pick_multiple_gpus(nb)
+
+
+@RunIf(min_gpus=2)
+@mock.patch("pytorch_lightning.trainer.connectors.accelerator_connector.pick_multiple_gpus", return_value=[1])
+def test_auto_select_gpus_with_devices_flag(pick_gpus_mock):
```
Comment (rohitgr7) [inline tests/trainer/properties/test_auto_gpu_select.py:66]:
```suggestion
def test_auto_select_gpus_with_devices_flag(_):
```

File: tests/trainer/properties/test_auto_gpu_select.py [hunk 2]
```diff
@@ -52,10 +52,30 @@ def test_pick_multiple_gpus(nb, expected_gpu_idxs, expected_error):
...
+
+    trainer = Trainer(auto_select_gpus=True, accelerator="gpu", devices=1)
+    assert trainer.num_devices == 1
+    assert trainer.device_ids == [1]
+
+
+@RunIf(min_gpus=2)
+@mock.patch("pytorch_lightning.trainer.connectors.accelerator_connector.pick_multiple_gpus", return_value=[1])
+def test_auto_select_gpus_with_gpus_flag(pick_gpus_mock):
```
Comment (rohitgr7) [inline tests/trainer/properties/test_auto_gpu_select.py:75]:
```suggestion
def test_auto_select_gpus_with_gpus_flag(_):
```

File: pytorch_lightning/tuner/auto_gpu_select.py [hunk 3]
```diff
@@ -22,15 +22,18 @@ def pick_multiple_gpus(nb: int) -> List[int]:
     """
     Raises:
         MisconfigurationException:
-            If ``gpus`` is set to 0, when ``auto_select_gpus=True``.
+            If ``gpus`` or ``devices`` is set to 0, when ``auto_select_gpus=True``.
     """
     if nb == 0:
         raise MisconfigurationException(
-            "auto_select_gpus=True, gpus=0 is not a valid configuration."
+            "auto_select_gpus=True, gpus/devices=0 is not a valid configuration."
```
Comment (rohitgr7) [inline pytorch_lightning/tuner/auto_gpu_select.py:29]: isn't this going to fail in your other PR? where you are checking devices = [], 0, '0'?
not sure which one comes first.

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 4]
```diff
@@ -408,7 +410,7 @@ def _check_device_config_and_set_final_flags(
         num_processes: Optional[int],
         gpus: Optional[Union[List[int], str, int]],
         ipus: Optional[int],
-        tpu_cores: Optional[Union[List[int], int]],
+        tpu_cores: Optional[Union[List[int], str, int]],
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:413]: str means "1" and "8"?

File: tests/trainer/properties/test_auto_gpu_select.py (hunk 2 above)
Comment (kaushikb11) [reply tests/trainer/properties/test_auto_gpu_select.py]: Updated!

File: pytorch_lightning/tuner/auto_gpu_select.py (hunk 3 above)
Comment (kaushikb11) [reply pytorch_lightning/tuner/auto_gpu_select.py:29]: One GPU test is failing for this PR, we would need the other one to be merged first!

Comment (kaushikb11) [reply pytorch_lightning/tuner/auto_gpu_select.py:29]: @rohitgr7

File: CHANGELOG.md [hunk 5]
```diff
@@ -40,7 +40,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Marked `swa_lrs` argument in `StochasticWeightAveraging` callback as required ([#12556](https://github.com/PyTorchLightning/pytorch-lightning/pull/12556))
 
 
--
+- Support `auto_select_gpus` with the accelerator and devices API ([#12608](https://github.com/PyTorchLightning/pytorch-lightning/pull/12608))
```
Comment (carmocca) [inline CHANGELOG.md:43]: If this goes into the 1.6.x milestone, we probably want to move it to "Fixed"

File: tests/trainer/properties/test_auto_gpu_select.py [hunk 6]
```diff
@@ -59,3 +60,23 @@ def test_pick_multiple_gpus(nb, expected_gpu_idxs, expected_error):
             pick_multiple_gpus(nb)
     else:
         assert expected_gpu_idxs == pick_multiple_gpus(nb)
+
+
+@RunIf(min_gpus=2)
```
Comment (carmocca) [inline tests/trainer/properties/test_auto_gpu_select.py:65]: Can we mock it?

File: tests/trainer/properties/test_auto_gpu_select.py [hunk 7]
```diff
@@ -59,3 +60,23 @@ def test_pick_multiple_gpus(nb, expected_gpu_idxs, expected_error):
...
+
+    trainer = Trainer(auto_select_gpus=True, accelerator="gpu", devices=1)
+    assert trainer.num_devices == 1
+    assert trainer.device_ids == [1]
+
+
+@RunIf(min_gpus=2)
+@mock.patch("pytorch_lightning.trainer.connectors.accelerator_connector.pick_multiple_gpus", return_value=[1])
+def test_auto_select_gpus_with_gpus_flag(_):
+
```
Comment (carmocca) [inline tests/trainer/properties/test_auto_gpu_select.py:77]: It woudl make sense to me to merge both tests. It will also be a tiny bit faster

```suggestion
```

**Review by carmocca — APPROVED** [review]

File: tests/trainer/properties/test_auto_gpu_select.py (hunk 6 above)
Comment (kaushikb11) [reply tests/trainer/properties/test_auto_gpu_select.py]: Done!

File: tests/trainer/properties/test_auto_gpu_select.py (hunk 7 above)
Comment (kaushikb11) [reply tests/trainer/properties/test_auto_gpu_select.py]: Done!

**Review by rohitgr7 — APPROVED** [review]
lgtm!

File: pytorch_lightning/tuner/auto_gpu_select.py [hunk 8]
```diff
@@ -22,15 +22,18 @@ def pick_multiple_gpus(nb: int) -> List[int]:
...
     if nb == 0:
         raise MisconfigurationException(
             "auto_select_gpus=True, gpus=0 is not a valid configuration."
             " Please select a valid number of GPU resources when using auto_select_gpus."
         )
 
-    nb = torch.cuda.device_count() if nb == -1 else nb
+    num_gpus = torch.cuda.device_count()
+    if nb > num_gpus:
+        raise MisconfigurationException(f"You requested {nb} gpus but your machine only has {num_gpus} gpus.")
```
Comment (awaelchli) [inline pytorch_lightning/tuner/auto_gpu_select.py:35]:
```suggestion
        raise MisconfigurationException(f"You requested {nb} GPUs but your machine only has {num_gpus} GPUs.")
```

**Review by carmocca — APPROVED** [review]
